File: backend/app/api/v1/llm.py
"""Demo endpoints proving the AI subsystem works end-to-end.

  POST /llm/chat   : one-shot completion
  POST /llm/stream : SSE stream of token deltas
  POST /llm/embed  : embedding vectors for a list of texts

These are open (no auth) for Phase 3 ease. Real agent endpoints in later
phases will require auth.
"""
from __future__ import annotations

import json
from typing import AsyncIterator

from fastapi import APIRouter
from fastapi.responses import StreamingResponse

from app.ai.embeddings.factory import get_embedder
from app.ai.llm.base import ChatMessage
from app.ai.llm.factory import get_llm
from app.schemas.llm import (
    ChatRequest,
    ChatResponse,
    EmbedRequest,
    EmbedResponse,
)

router = APIRouter()


@router.post("/chat", response_model=ChatResponse, summary="One-shot LLM completion")
async def chat(req: ChatRequest) -> ChatResponse:
    llm = get_llm()
    messages = [ChatMessage(role=m.role, content=m.content) for m in req.messages]
    reply = await llm.chat(
        messages,
        temperature=req.temperature,
        max_tokens=req.max_tokens,
    )
    return ChatResponse(
        provider=llm.provider_name,
        model=llm.model_name,
        content=reply,
    )


@router.post("/stream", summary="SSE-streamed LLM completion (text/event-stream)")
async def chat_stream(req: ChatRequest) -> StreamingResponse:
    llm = get_llm()
    messages = [ChatMessage(role=m.role, content=m.content) for m in req.messages]

    async def _sse() -> AsyncIterator[bytes]:
        async for delta in llm.stream(
            messages, temperature=req.temperature, max_tokens=req.max_tokens
        ):
            payload = json.dumps({"delta": delta}, ensure_ascii=False)
            yield f"data: {payload}\n\n".encode("utf-8")
        yield b"data: [DONE]\n\n"

    return StreamingResponse(_sse(), media_type="text/event-stream")


@router.post("/embed", response_model=EmbedResponse, summary="Embed texts to vectors")
async def embed(req: EmbedRequest) -> EmbedResponse:
    embedder = get_embedder()
    vectors = await embedder.embed_documents(req.texts)
    return EmbedResponse(
        model=embedder.model_name,
        dimension=embedder.dimension,
        embeddings=vectors,
    )

File: backend-ts/src/routes/llm.ts
/**
 * Public LLM endpoints (no auth — mirror the Python backend).
 *
 *   POST /llm/chat    one-shot chat completion
 *   POST /llm/embed   embed a list of texts
 *   POST /llm/stream  SSE-style streaming chat
 */
import type { FastifyInstance } from "fastify";
import { z } from "zod";
import { getLLM } from "../ai/llm/factory.js";
import { embed } from "../ai/embeddings/embedder.js";

const chatSchema = z.object({
  messages: z
    .array(
      z.object({
        role: z.enum(["system", "user", "assistant"]),
        content: z.string(),
      }),
    )
    .min(1),
  temperature: z.number().min(0).max(2).optional(),
  max_tokens: z.number().int().positive().optional(),
  json_mode: z.boolean().optional(),
});

const embedSchema = z.object({
  texts: z.array(z.string()).min(1).max(256),
});

export default async function llmRoutes(app: FastifyInstance) {
  app.post("/chat", async (req, reply) => {
    const parsed = chatSchema.safeParse(req.body);
    if (!parsed.success) return reply.code(422).send({ detail: parsed.error.flatten() });
    const llm = getLLM();
    return llm.chat(parsed.data.messages, {
      temperature: parsed.data.temperature,
      maxTokens: parsed.data.max_tokens,
      jsonMode: parsed.data.json_mode,
    });
  });

  app.post("/embed", async (req, reply) => {
    const parsed = embedSchema.safeParse(req.body);
    if (!parsed.success) return reply.code(422).send({ detail: parsed.error.flatten() });
    return embed(parsed.data.texts);
  });

  app.post("/stream", async (req, reply) => {
    const parsed = chatSchema.safeParse(req.body);
    if (!parsed.success) return reply.code(422).send({ detail: parsed.error.flatten() });
    // Hand control of the raw socket to us so Fastify doesn't try to send a body.
    reply.hijack();
    reply.raw.setHeader("Content-Type", "text/event-stream; charset=utf-8");
    reply.raw.setHeader("Cache-Control", "no-cache");
    reply.raw.setHeader("Connection", "keep-alive");
    reply.raw.flushHeaders?.();

    const llm = getLLM();
    try {
      for await (const chunk of llm.stream(parsed.data.messages, {
        temperature: parsed.data.temperature,
        maxTokens: parsed.data.max_tokens,
      })) {
        reply.raw.write(`data: ${JSON.stringify({ delta: chunk })}\n\n`);
      }
      reply.raw.write("data: [DONE]\n\n");
    } catch (e) {
      reply.raw.write(`data: ${JSON.stringify({ error: (e as Error).message })}\n\n`);
    } finally {
      reply.raw.end();
    }
  });
}
